pkg/crcontrollers/storagegroup/storagegroupcontroller.go

```diff
@@ -0,0 +1,586 @@
+		// TODO need to restrict storage group update event handling
+		return true
+	}
+	if newDrive, ok = new.(*drivecrd.Drive); ok {
+		return filterDrive(oldDrive, newDrive)
+	}
+	return false
+}
+
+func filterDrive(old *drivecrd.Drive, new *drivecrd.Drive) bool {
+	return old.Labels[apiV1.StorageGroupLabelKey] != new.Labels[apiV1.StorageGroupLabelKey]
+}
+
+// Reconcile reconciles StorageGroup custom resources
+func (c *Controller) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	ctx, cancelFn := context.WithTimeout(ctx, contextTimeoutSeconds*time.Second)
+	defer cancelFn()
+
+	// read name
+	name := req.Name
+	// customize logging
+	log := c.log.WithFields(logrus.Fields{"method": "Reconcile", "name": name})
+
+	drive := &drivecrd.Drive{}
+	if err := c.client.ReadCR(ctx, name, "", drive); err == nil {
+		return c.syncDriveStorageGroupLabel(ctx, drive)
+	}
+
+	storageGroup := &sgcrd.StorageGroup{}
+	if err := c.client.ReadCR(ctx, name, "", storageGroup); err != nil {
+		log.Warningf("Failed to read StorageGroup %s", name)
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	log.Debugf("Reconcile StorageGroup: %v", storageGroup)
+
+	// StorageGroup Deletion request
+	if !storageGroup.DeletionTimestamp.IsZero() {
+		return c.handleStorageGroupDeletion(ctx, log, storageGroup)
+	}
+
+	if !util.ContainsString(storageGroup.Finalizers, sgFinalizer) {
+		// append finalizer
+		log.Debugf("Appending finalizer for StorageGroup")
+		storageGroup.Finalizers = append(storageGroup.Finalizers, sgFinalizer)
+		if err := c.client.UpdateCR(ctx, storageGroup); err != nil {
+			log.Errorf("Unable to append finalizer %s to StorageGroup, error: %v.", sgFinalizer, err)
+			return ctrl.Result{Requeue: true}, err
+		}
+	}
+
+	if storageGroup.Annotations == nil {
+		storageGroup.Annotations = map[string]string{}
+	}
+	sgStatus, ok := storageGroup.Annotations[sgTempStatusAnnotationKey]
+	if !ok {
+		if !c.isStorageGroupValid(log, storageGroup) {
+			storageGroup.Annotations[sgTempStatusAnnotationKey] = apiV1.Failed
+			if err := c.client.UpdateCR(ctx, storageGroup); err != nil {
+				log.Errorf("Unable to update StorageGroup status, error: %v.", err)
+				return ctrl.Result{Requeue: true}, err
+			}
+			return ctrl.Result{}, nil
+		}
+		// Pass storage group valiation, change to CREATING status
+		sgStatus = apiV1.Creating
+		storageGroup.Annotations[sgTempStatusAnnotationKey] = apiV1.Creating
+		if err := c.client.UpdateCR(ctx, storageGroup); err != nil {
+			log.Errorf("Unable to update StorageGroup status, error: %v.", err)
+			return ctrl.Result{Requeue: true}, err
+		}
+	}
+
+	if sgStatus == apiV1.Creating {
+		return c.handleStorageGroupCreation(ctx, log, storageGroup)
+	}
+
+	return ctrl.Result{}, nil
+}
+
+// combine the following similar funcs
+func (c *Controller) removeACStorageGroupLabel(ctx context.Context, log *logrus.Entry, ac *accrd.AvailableCapacity) error {
+	delete(ac.Labels, apiV1.StorageGroupLabelKey)
```

Comment:
is any chance ac.Labels is nil ?

Reply:
In most cases, AC will have CSI default labels. But I will handle the case you mentioned.